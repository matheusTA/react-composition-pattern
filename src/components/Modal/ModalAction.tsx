import { twMerge } from 'tailwind-merge'

const Colors = {
  success: 'bg-green-500',
  error: 'bg-red-500',
  warning: 'bg-yellow-500',
  info: 'bg-blue-500',
  none: 'bg-transparent',
} as const

interface ModalActionProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  color?: 'success' | 'warning' | 'error' | 'info' | 'none'
}

export function ModalAction({
  color = 'none',
  className,
  onClick,
  children,
  ...restProps
}: ModalActionProps) {
  const colorButton = Colors[color]

  return (
    <button
      {...restProps}
      onClick={onClick}
      className={twMerge(
        'rounded-md px-4 py-1 text-white',
        className,
        colorButton,
      )}
    >
      {children}
    </button>
  )
}
