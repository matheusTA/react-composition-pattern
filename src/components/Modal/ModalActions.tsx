interface ModalActionsProps {
  children: React.ReactNode
}

export function ModalActions({ children }: ModalActionsProps) {
  return (
    <footer className="flex w-full justify-end gap-2 p-3">{children}</footer>
  )
}
