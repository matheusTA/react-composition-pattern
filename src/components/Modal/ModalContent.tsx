interface ModalContentProps {
  children: React.ReactNode
}

export function ModalContent({ children }: ModalContentProps) {
  return (
    <div className="flex flex-1 items-center justify-center p-3">
      {children}
    </div>
  )
}
