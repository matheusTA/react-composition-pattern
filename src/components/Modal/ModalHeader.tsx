interface ModalHeaderProps {
  children: React.ReactNode
}

export function ModalHeader({ children }: ModalHeaderProps) {
  return (
    <header className="bg-slate-200">
      <div className="flex items-center justify-between p-3">{children}</div>
    </header>
  )
}
