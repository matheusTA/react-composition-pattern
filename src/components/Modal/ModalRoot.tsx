interface ModalRootProps {
  isOpen: boolean
  children: React.ReactNode
}

export function ModalRoot({ isOpen, children }: ModalRootProps) {
  return (
    <>
      {isOpen && (
        <div className="fixed inset-0 z-40 bg-black/25 text-gray-700">
          <div className="flex h-full w-full items-center justify-center">
            <div className="absolute z-50 mx-auto my-0 flex h-80 w-[480px] flex-col overflow-hidden rounded-xl bg-white shadow-2xl">
              {children}
            </div>
          </div>
        </div>
      )}
    </>
  )
}
