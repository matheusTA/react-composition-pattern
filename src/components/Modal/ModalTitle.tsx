type ModalTitleProps = {
  children: React.ReactNode
}

export function ModalTitle({ children }: ModalTitleProps) {
  return <p className="text-md font-bold text-gray-600">{children}</p>
}
