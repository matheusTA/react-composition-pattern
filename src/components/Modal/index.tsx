import { ModalRoot } from './ModalRoot'
import { ModalHeader } from './ModalHeader'
import { ModalTitle } from './ModalTitle'
import { ModalContent } from './ModalContent'
import { ModalActions } from './ModalActions'
import { ModalAction } from './ModalAction'

export const Modal = {
  Root: ModalRoot,
  Header: ModalHeader,
  Title: ModalTitle,
  Content: ModalContent,
  Actions: ModalActions,
  Action: ModalAction,
}
