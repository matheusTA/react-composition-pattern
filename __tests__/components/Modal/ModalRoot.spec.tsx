import { render } from '@testing-library/react'
import { ModalRoot } from '@/components/Modal/ModalRoot'

import '@testing-library/jest-dom'

describe('ModalRoot component', () => {
  it('renders ModalRoot unchanged', () => {
    const isOpen = true
    const childElement = <p>Modal Content</p>

    const { container } = render(
      <ModalRoot isOpen={isOpen}>{childElement}</ModalRoot>,
    )

    expect(container).toMatchSnapshot()
  })

  it('renders modal content when isOpen is true', () => {
    const isOpen = true
    const childElement = <p>Modal Content</p>

    const { getByText } = render(
      <ModalRoot isOpen={isOpen}>{childElement}</ModalRoot>,
    )

    expect(getByText('Modal Content')).toBeInTheDocument()
  })

  it('does not render modal content when isOpen is false', () => {
    const isOpen = false
    const childElement = <p>Modal Content</p>

    const { queryByText } = render(
      <ModalRoot isOpen={isOpen}>{childElement}</ModalRoot>,
    )

    expect(queryByText('Modal Content')).toBeNull()
  })

  it('applies correct CSS classes when modal is open', () => {
    const isOpen = true
    const childElement = <p>Modal Content</p>

    const { container } = render(
      <ModalRoot isOpen={isOpen}>{childElement}</ModalRoot>,
    )
    const modalRootElement = container.firstChild

    expect(modalRootElement).toHaveClass(
      'fixed inset-0 z-40 bg-black/25 text-gray-700',
    )
  })
})
