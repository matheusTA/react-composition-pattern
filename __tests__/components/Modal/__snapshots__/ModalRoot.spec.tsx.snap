// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalRoot component renders ModalRoot unchanged 1`] = `
<div>
  <div
    class="fixed inset-0 z-40 bg-black/25 text-gray-700"
  >
    <div
      class="flex h-full w-full items-center justify-center"
    >
      <div
        class="absolute z-50 mx-auto my-0 flex h-80 w-[480px] flex-col overflow-hidden rounded-xl bg-white shadow-2xl"
      >
        <p>
          Modal Content
        </p>
      </div>
    </div>
  </div>
</div>
`;
